src/main/java/dev/meyi/noscamspam/checks/UserCheck.java

```diff
@@ -13,40 +18,51 @@
-      StatsResponse statsResponse = NoScamSpam.gson
-          .fromJson(IOUtils
-              .toString(stream,
-                  StandardCharsets.UTF_8), StatsResponse.class);
-
-      if (statsResponse.success) {
-        if (statsResponse.player.networkExp > NoScamSpam.config.hypixelExperience &&
-            statsResponse.player.achievements.getSkillAverage() > NoScamSpam.config.skillAverage &&
-            statsResponse.player.achievements.skyblock_dungeoneer
-                > NoScamSpam.config.catacombsLevel) {
-          return false;
-        } else {
-          if (!NoScamSpam.blacklist.contains(username)) {
-            NoScamSpam.blacklist.add(username);
-          }
-          return true;
+    public static void shouldBlockUser(String username, ClientChatReceivedEvent event) {
+        if (NoScamSpam.config.whitelist.contains(username) || NoScamSpam.cachedAllowed.contains(username))
+            return;
+        if (NoScamSpam.blacklist.contains(username)) {
+            event.setCanceled(true);
+            return;
         }
-      }
-    } catch (IOException e) {
-      e.printStackTrace();
-    }
 
-    return !NoScamSpam.config.showOnError;
-  }
+        event.setCanceled(true);
+        Thread fetchThread = new Thread(() -> {
+            try (CloseableHttpResponse response = httpClient.execute(new HttpGet(
+                    "https://api.hypixel.net/player?key=" + NoScamSpam.config.apiKey + "&name=" + username));
+                 InputStream stream = response.getEntity().getContent()) {
+                StatsResponse statsResponse = NoScamSpam.gson
+                        .fromJson(IOUtils
+                                .toString(stream,
+                                        StandardCharsets.UTF_8), StatsResponse.class);
+
+                if (statsResponse.success) {
+                    if (statsResponse.player.networkExp > NoScamSpam.config.hypixelExperience &&
+                            statsResponse.player.achievements.getSkillAverage() > NoScamSpam.config.skillAverage &&
+                            statsResponse.player.achievements.skyblock_dungeoneer
+                                    > NoScamSpam.config.catacombsLevel) {
+                        Minecraft.getMinecraft().thePlayer.addChatMessage(event.message);
+                        NoScamSpam.cachedAllowed.add(username);
+                        ClientChatReceivedEvent newEvent = new ClientChatReceivedEvent((byte) 0, event.message);
+                        MinecraftForge.EVENT_BUS.post(newEvent);
+                    } else
+                        NoScamSpam.blacklist.add(username);
+                    Thread.currentThread().interrupt();
+                    return;
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+
+            if (NoScamSpam.config.showOnError) {
+                Minecraft.getMinecraft().thePlayer.addChatMessage(event.message);
```

Comment:
Why would we `addChatMessage` instead of just letting the event go through as if this mod didn't exist? I haven't encountered any issues with it as is.

Reply:
because forge events do block the game loop